File: Bwl.Framework.Windows/src/Settings/Gui/SettingsDialog.vb
Imports System.Windows.Forms

Public Class SettingsDialog
    Implements ISettingsForm

    Private storage As ISettingsStorage
    Private Event SettingsFormClosed() Implements ISettingsForm.SettingsFormClosed

    Public Sub ShowSettings(newStorage As ISettingsStorage) Implements ISettingsForm.ShowSettings
        storage = newStorage
        FillTree(storage)
        Me.Text = "Настройки " & storage.CategoryName
    End Sub

    Public Shadows Sub ShowDialog() Implements ISettingsForm.ShowDialog
        MyBase.ShowDialog()
    End Sub

    Public Shadows Sub Show() Implements ISettingsForm.Show
        MyBase.Show()
    End Sub

    Private Sub SettingsDialog_FormClosed(sender As Object, e As FormClosedEventArgs) Handles Me.FormClosed
        RaiseEvent SettingsFormClosed()
    End Sub

    Private Sub FillTree(storage As ISettingsStorage)
        list.Nodes.Clear()
        Dim nodeList As New TreeNode
        FillTreeRecursive(nodeList, storage)
        For Each node As TreeNode In nodeList.Nodes
            list.Nodes.Add(node)
        Next
    End Sub
    Private Sub FillTreeRecursive(node As TreeNode, storage As ISettingsStorage)
        For Each childStorage As ISettingsStorage In storage.ChildStorages
            Dim newNode As New TreeNode
            newNode.Text = childStorage.CategoryName
            If childStorage.FriendlyCategoryName.Length > 0 Then
                newNode.Text = childStorage.FriendlyCategoryName
            End If
            newNode.ToolTipText = ""
            newNode.ImageIndex = 0
            newNode.SelectedImageIndex = 0
            FillTreeRecursive(newNode, childStorage)
            node.Nodes.Add(newNode)
        Next

        For Each childSetting As SettingOnStorage In storage.GetSettings
            Dim newNode As New TreeNode
            newNode.ImageIndex = 1
            newNode.SelectedImageIndex = 1
            Dim nameText As String = childSetting.Name
            If childSetting.FriendlyName.Length > 0 Then
                nameText = childSetting.FriendlyName
            End If
            Dim val = childSetting.ValueAsString
            If (val.Length > 30) Then
                'val = val.Substring(0, 30)
            End If
            newNode.Text = nameText & ": " & val

            newNode.ToolTipText = childSetting.Description
            newNode.Tag = childSetting
            node.Nodes.Add(newNode)
        Next
    End Sub

    Private Sub list_AfterSelect(sender As System.Object, e As System.Windows.Forms.TreeViewEventArgs) Handles list.AfterSelect
        If list.SelectedNode.Tag IsNot Nothing Then
            settingView.AssignedSetting = DirectCast(list.SelectedNode.Tag, SettingOnStorage)
        End If
    End Sub

    Private Sub settingView_SettingValueChanged() Handles settingView.SettingValueChanged
        If Not Me.IsDisposed Then
            If list.SelectedNode.Tag IsNot Nothing Then
                Dim setting As SettingOnStorage = DirectCast(list.SelectedNode.Tag, SettingOnStorage)
                Dim nameText As String = setting.Name
                If setting.FriendlyName.Length > 0 Then
                    nameText = setting.FriendlyName
                End If
                Dim val = setting.ValueAsString
                If (val.Length > 30) Then
                    'val = val.Substring(0, 30)
                End If
                list.SelectedNode.Text = nameText & ": " & val & " [*]"
            End If
        End If
    End Sub

    Private _bigFieldEnabled As Boolean = False

    Private Sub settingView_setBiggerField(value As Boolean) Handles settingView.SetBiggerField
        If _bigFieldEnabled = value Then Return
        Dim sizeChange = 65
        If value Then
            list.Height -= sizeChange
            settingView.Top -= sizeChange
            settingView.Height += sizeChange
        Else
            list.Height += sizeChange
            settingView.Top += sizeChange
            settingView.Height -= sizeChange
        End If
        _bigFieldEnabled = value
    End Sub
End Class

File: Framework/Framework/Bwl.Framework.Windows/Settings/Gui/SettingsDialog.cs
using System;
using System.Runtime.Versioning;
using System.Windows.Forms;

namespace Bwl.Framework.Windows
{

    [SupportedOSPlatform("windows")]
    public partial class SettingsDialog : ISettingsForm
    {
        private ISettingsStorage storage;
        private event EventHandler FormClosed;
        private event EventHandler Load;

        string IUIWindow.Text { get => base.Text; set => base.Text = value; }

        public SettingsDialog()
        {
            InitializeComponent();
        }

        event EventHandler IUIWindow.Load
        {
            add
            {
                Load += value;
            }
            remove
            {
                Load -= value;
            }
        }

        event EventHandler IUIWindow.FormClosed
        {
            add
            {
                FormClosed += value;
            }
            remove
            {
                FormClosed -= value;
            }
        }

        void ISettingsForm.ShowSettings(ISettingsStorage settingsStorage)
        {
            storage = settingsStorage;
            FillTree(storage);
            Text = "Настройки " + storage.CategoryName;
        }

        void IUIWindow.ShowDialog(object invokeForm)
        {
            ShowDialog((Form)invokeForm);
        }

        void IUIWindow.Show()
        {
            Show();
        }

        void IUIWindow.Close()
        {
            Close();
        }

        void IUIWindow.Invoke(Action action)
        {
            Invoke(action);
        }

        private void SettingsDialog_FormClosed(object sender, FormClosedEventArgs e)
        {
            FormClosed?.Invoke(sender, e);
        }

        private void SettingsDialog_FormLoad(object sender, EventArgs e)
        {
            Load?.Invoke(sender, e);
        }

        private void FillTree(ISettingsStorage storage)
        {
            list.Nodes.Clear();
            var nodeList = new TreeNode();
            FillTreeRecursive(nodeList, storage);
            foreach (TreeNode node in nodeList.Nodes)
                list.Nodes.Add(node);
        }
        private void FillTreeRecursive(TreeNode node, ISettingsStorage storage)
        {
            foreach (ISettingsStorage childStorage in storage.ChildStorages)
            {
                var newNode = new TreeNode();
                newNode.Text = childStorage.CategoryName;
                if (childStorage.FriendlyCategoryName.Length > 0)
                {
                    newNode.Text = childStorage.FriendlyCategoryName;
                }
                newNode.ToolTipText = "";
                newNode.ImageIndex = 0;
                newNode.SelectedImageIndex = 0;
                FillTreeRecursive(newNode, childStorage);
                node.Nodes.Add(newNode);
            }

            foreach (SettingOnStorage childSetting in storage.GetSettings())
            {
                var newNode = new TreeNode();
                newNode.ImageIndex = 1;
                newNode.SelectedImageIndex = 1;
                string nameText = childSetting.Name;
                if (childSetting.FriendlyName.Length > 0)
                {
                    nameText = childSetting.FriendlyName;
                }
                string val = childSetting.ValueAsString;
                if (val.Length > 30)
                {
                    // val = val.Substring(0, 30)
                }
                newNode.Text = nameText + ": " + val;
                newNode.ToolTipText = childSetting.Description;
                newNode.Tag = childSetting;
                node.Nodes.Add(newNode);
            }
        }

        private void list_AfterSelect(object sender, TreeViewEventArgs e)
        {
            if (list.SelectedNode.Tag is not null)
            {
                settingView.AssignedSetting = (SettingOnStorage)list.SelectedNode.Tag;
            }
        }

        private void settingView_SettingValueChanged()
        {
            if (!IsDisposed)
            {
                if (list.SelectedNode.Tag is not null)
                {
                    SettingOnStorage setting = (SettingOnStorage)list.SelectedNode.Tag;
                    string nameText = setting.Name;
                    if (setting.FriendlyName.Length > 0)
                    {
                        nameText = setting.FriendlyName;
                    }
                    string val = setting.ValueAsString;
                    if (val.Length > 30)
                    {
                        // val = val.Substring(0, 30)
                    }
                    list.SelectedNode.Text = nameText + ": " + val + " [*]";
                }
            }
        }

        private bool _bigFieldEnabled = false;

        private void settingView_setBiggerField(bool value)
        {
            if (_bigFieldEnabled == value)
                return;
            int sizeChange = 65;
            if (value)
            {
                list.Height -= sizeChange;
                settingView.Top -= sizeChange;
                settingView.Height += sizeChange;
            }
            else
            {
                list.Height += sizeChange;
                settingView.Top += sizeChange;
                settingView.Height -= sizeChange;
            }
            _bigFieldEnabled = value;
        }
    }
}